Fix megacheck issues

mjtrangoni opened the pull request:
Hi @torkelo @bergquist,

related to #10381, I fixed many *megacheck* issues.

I will be fixing this in a separate PR,
```
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/tsdb/influxdb/influxdb.go:106:3:warning: the surrounding loop is unconditionally terminated (SA4004) (megacheck)
```

Do you agree adding [megacheck](https://github.com/dominikh/go-tools/tree/master/cmd/megacheck) to CircleCI?

mjtrangoni committed:
pkg/plugins/plugins.go: remove ineffective break statement.

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/plugins/plugins.go:124:4:warning: ineffective break statement. Did you mean to break out of the outer loop? (SA4011) (megacheck)

mjtrangoni committed:
pkg/middleware/middleware.go: Fix empty branch warning.

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/middleware/middleware.go:46:3:warning: empty branch (SA9003) (megacheck)

mjtrangoni committed:
pkg/services/alerting/reader.go: Fix should use for range instead of for { select {} }.

$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/services/alerting/reader.go:37:2:warning: should use for range instead of for { select {} } (S1000) (megacheck)

mjtrangoni committed:
scripts/build/publish.go: Fix warning on err variable.

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
scripts/build/publish.go:126:48:warning: argument err is overwritten before first use (SA4009) (megacheck)

bergquist commented on pkg/middleware/middleware.go:
```diff
@@ -49,6 +49,7 @@ func GetContextHandler() macaron.Handler {
 			initContextWithAuthProxy(ctx, orgId) ||
 			initContextWithUserSessionCookie(ctx, orgId) ||
 			initContextWithAnonymousUser(ctx) {
+			log.Debug2("Context initialized")
```
This logline is too verbose. I think we should rewrite the chain rather than add dummy logging.

mjtrangoni commented:
@bergquist PTAL

torkelo commented on pkg/tsdb/influxdb/influxdb.go:
```diff
@@ -96,17 +96,22 @@ func (e *InfluxDBExecutor) Query(ctx context.Context, dsInfo *models.DataSource,
 	}
-
 	return nil, fmt.Errorf("query request contains no queries")
+found:
```
this seems like a very unconventional change for the worse (in terms of readability).  must be a better way?

mjtrangoni replied:
@torkelo yes, could you explain me this a little bit? I do not get why there is a for loop here.

torkelo replied:
Good question, I think the model supports multiple queries but right now this is only used from alerting so we only needed to support batch executing 1 query at a time. So it just takes the first query.

mjtrangoni replied:
@torkelo ok. Should I remove that for loop and make a comment about this for future implementations?

torkelo replied:
What is the problem with the loop? Is a length check index lookup better ?

mjtrangoni replied:
@torkelo this for loop is not needed, because it does only *one* loop, and as the warning say, it terminates unconditionally (no brake, continue, goto, etc). I will change this to parse the first element, and returns what corresponds to.

mjtrangoni replied:
@torkelo Done. PTAL.

mjtrangoni commented on pkg/middleware/middleware.go:
```diff
@@ -43,12 +43,13 @@ func GetContextHandler() macaron.Handler {
-			initContextWithAuthProxy(ctx, orgId) ||
-			initContextWithUserSessionCookie(ctx, orgId) ||
-			initContextWithAnonymousUser(ctx) {
+		switch {
```
@torkelo so is it ok for you? I double-checked it, and it respects the order.

mjtrangoni committed:
pkg/login/ldap.go: Fix warning comparison to bool constant

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/login/ldap.go:188:40:warning: should omit comparison to bool constant, can be simplified to !*extUser.IsGrafanaAdmin (S1002) (megacheck)

mjtrangoni commented:
@torkelo @bergquist I will be fixing the 3 new warnings tomorrow.

mjtrangoni committed:
pkg/tsdb/graphite/graphite.go: Fix regular expression does not contain any meta characters.

I found this article benchmarking Replace vs Regexp,
https://medium.com/codezillas/golang-replace-vs-regexp-de4e48482f53

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/tsdb/graphite/graphite.go:167:28:warning: regular expression does not contain any meta characters (SA6004) (megacheck)
pkg/tsdb/graphite/graphite.go:172:28:warning: regular expression does not contain any meta characters (SA6004) (megacheck)

mjtrangoni committed:
pkg/tsdb/stackdriver/stackdriver.go: Fix regular expression does not contain any meta characters.

See,
$ gometalinter --vendor --deadline 10m --disable-all --enable=megacheck ./...
pkg/tsdb/stackdriver/stackdriver.go:171:26:warning: regular expression does not contain any meta characters (SA6004) (megacheck)

mjtrangoni commented:
@torkelo It is fixed now, and rebased against master. PTAL.

mjtrangoni commented:
@torkelo @bergquist Ping? Should I drop something to get this merged? I am going on vacations tomorrow.

torkelo commented on pkg/tsdb/influxdb/influxdb.go:
```diff
@@ -96,17 +96,13 @@ func (e *InfluxDBExecutor) Query(ctx context.Context, dsInfo *models.DataSource,
-		return query, nil
+	// The model supports multiple queries, but right now this is only used from
+	// alerting so we only needed to support batch executing 1 query at a time.
+	query, err := e.QueryParser.Parse(queries[0].Model, dsInfo)
```
This still feels like a change for the worse. The old code would not throw index out of range panic for empty queries arrays where with this change it would. If you remove the range loop you need to add a length check and return the error that the old code did when no queries found

mjtrangoni replied:
@torkelo  No problem. I dropped this change, so that I will be fixing it later on, adding Megacheck to the CI.

mjtrangoni commented:
@torkelo PTAL. I hope this can be merged now.

torkelo commented:
Thanks for contributing!
